content/blog/2022/2022-08-03-feature-flags-in-nodejs-react-client-with-launchdarkly.md

```diff
@@ -0,0 +1,305 @@
+import React, { Component } from 'react';
+import './App.css';
+import * as LDClient from 'launchdarkly-js-client-sdk';
+
+class App extends Component {
+  constructor() {
+    super()
+    this.state = {
+      selectedSortOrder: null,
+      users: [
+        { name: 'John Doe', added: new Date('2022-7-27') },
+        { name: 'Allen Witt', added: new Date('2022-6-30') },
+        { name: 'Cheryl Strong', added: new Date('2022-7-02') },
+        { name: 'Marty Byrde', added: new Date('2022-5-03') },
+        { name: 'Wendy Byrde', added: new Date('2022-6-03') },
+      ]
+    }
+  }
+  componentDidMount() {
+    const user = {
+      // UI based user
+      key: 'user_a'
+    }
+    // SDK requires Client-side ID for UI call
+    this.ldclient = LDClient.initialize('62e9289ade464c10d842c2b3', user);
+    this.ldclient.on('ready', this.onLaunchDarklyUpdated.bind(this));
+    this.ldclient.on('change', this.onLaunchDarklyUpdated.bind(this));
+  }
+  onLaunchDarklyUpdated() {
+    this.setState({
+      featureFlags: {
+        hideUser: this.ldclient.variation('hidden-user', '')
+      }
+    })
+  }
+  render() {
+    if (!this.state.featureFlags) {
+      return <div className="App">Loading....</div>
+    }
+    
+    return (
+      <div className="App">
+        <div style ={{ fontWeight: 'bold' }}><h1>Users List</h1></div>
+        <ul>
+          {this.state.users.map(user =>
+             <div>{ this.state.featureFlags.hideUser === 'John Doe'
+              && user.name === 'John Doe' ? '*********' : user.name }</div>
+          )}
+        </ul>
+      </div>
+    );
+  }
+}
+
+export default App;
+```
+
+To execute code only when the LaunchDarkly client is ready, we have two mechanisms: an *event* or a *promise*.
+
+With `this.ldclient.on('ready', ...)`, we subscribe to the `ready` event which will fire once the LaunchDarkly client has received the state of all feature flags from the server.
+
+With `this.ldclient.on('change', ...)`, we subscribe to the `change` event which will fire once the LaunchDarkly client has received the state of all feature flags that has got updated or changed in the LaunchDarkly UI.
+
+For the promise mechanism, the SDK supports two methods: `waitUntilReady()` and `waitForInitialization()`. The behavior of `waitUntilReady()` is equivalent to the `ready` event. The promise resolves when the client receives its initial flag data. As with all promises, you can either use `.then()` to provide a callback, or use `await` if you are writing asynchronous code. The other method that returns a promise, `waitForInitialization()`, is similar to `waitUntilReady()` except that it also tells you if initialization fails by rejecting the promise.
+
+Next we can run the following command to start the UI:
+
+```bash
+npm start
+```
+
+When the user is not set in the LaunchDarkly UI, our React UI looks like below:
+
+{{% image alt="React User List" src="images/posts/nodejs-frontend-launchdarkly/React_User_List_simple.png" %}}
```

Comment:
These images are a lot of white and a tiny user list in the middle. Can you zoom in to the list, please? Also in the other screen shots.

Reply:
Updated the image by zooming it.